src/nvim/undo.c

```diff
@@ -2035,83 +2033,97 @@ void undo_time(long step, int sec, int file, int absolute)
         uhp = curbuf->b_u_newhead;
       else
         uhp = uhp->uh_next.ptr;
-      if (uhp == NULL || uhp->uh_walk != mark
-          || (uhp->uh_seq == target && !above))
+      if (uhp == NULL
+          || (target > 0 && uhp->uh_walk != mark)
+          || (uhp->uh_seq == target && !above)) {
         break;
+      }
       curbuf->b_u_curhead = uhp;
       u_undoredo(true, true);
-      uhp->uh_walk = nomark;            // don't go back down here
+      if (target > 0) {
+        uhp->uh_walk = nomark;          // don't go back down here
+      }
     }
 
-    /*
-     * And now go down the tree (redo), branching off where needed.
-     */
-    while (!got_int) {
-      /* Do the change warning now, for the same reason as above. */
-      change_warning(0);
-
-      uhp = curbuf->b_u_curhead;
-      if (uhp == NULL)
-        break;
+    // When back to origin, redo is not needed.
+    if (target > 0) {
+      // And now go down the tree (redo), branching off where needed.
+      while (!got_int) {
```

Comment:
This entire while statement and its block scope are indented and include lint changes.